aptos-move/move-examples/defi/aptoswap/core/sources/liquidity_pool.move

```diff
@@ -0,0 +1,442 @@
+    use aptos_framework::object::{Self, ExtendRef, Object, generate_signer_for_extending, generate_extend_ref};
+    use aptos_framework::fungible_asset;
+    use aptos_framework::primary_fungible_store;
+    use aptos_std::comparator;
+    use aptos_std::math128;
+    use aptos_std::math64;
+    use aptos_std::smart_table::{Self, SmartTable};
+    use aptos_std::string_utils;
+    use std::bcs;
+    use std::error;
+    use std::option::{Self, Option};
+    use std::string::{Self, String};
+    use std::vector;
+
+    const MINIMAL_LIQUIDITY: u64 = 1000;
+
+    // fee is basis point based so the scale is 10000.
+    const FEE_SCALE: u64 = 10000;
+
+    /// A meaningful swap pair must be different assets.
+    const ERR_SWAP_PAIR_CANNOT_BE_THE_SAME: u64 = 1;
+    /// The liquidity pool already exists.
+    const ERR_POOL_EXISTENCE: u64 = 2;
+    /// The minted liquidity is too small.
+    const ERR_NOT_ENOUGH_LIQUIDITY: u64 = 3;
+    /// The number of lp tokens to burn is too small to redeem both assets.
+    const ERR_NOT_ENOUGH_LP_TOKEN_TO_BURN: u64 = 4;
+    /// The number of assets to swap cannot be zero.
+    const ERR_ZERO_COIN_TO_SWAP: u64 = 5;
+    /// The swap function arguments are not valid because it will result in lower liquidity after the swap.
+    const ERR_INCORRECT_SWAP_DUE_TO_REDUCED_LIQUIDITY: u64 = 6;
+
+    struct LiquidityPool has key {
+        extend_ref: ExtendRef,
+        /// a vector of length 2, i.e. an object pair, whose addresses must be lexicographically ordered.
+        store_pair: vector<Object<FungibleStore>>,
+        lp_token_mint_ref: fungible_asset::MintRef,
+        lp_token_burn_ref: fungible_asset::BurnRef,
+        // basis points
+        fee: u64,
+    }
+
+    /// The onchain index of the liquidity pools of different pairs
+    struct LiquidityPoolManager has key {
+        extend_ref: ExtendRef,
+        map: SmartTable<vector<Object<Metadata>>, Object<LiquidityPool>>
+    }
+
+    /// Intialize the module in pool manager object
+    fun init_module(admin: &signer) {
+        let extend_ref = aptoswap::init::retrieve_extend_ref(admin);
+        let pool_manager_obj_signer = object::generate_signer_for_extending(&extend_ref);
+        move_to(&pool_manager_obj_signer, LiquidityPoolManager {
+            extend_ref,
+            map: smart_table::new()
+        })
+    }
+
+    /// Create liquidity pool `X`/`Y`.
+    public fun create(acc: &signer, x: Object<Metadata>, y: Object<Metadata>) acquires LiquidityPoolManager {
+        if (!is_valid_pair(x, y)) {
+            return create(acc, y, x)
+        };
+        assert!(option::is_none(&get_liquidity_pool(x, y)), error::already_exists(ERR_POOL_EXISTENCE));
+
+        let (lp_token_name, lp_token_symbol) = lp_token_name_and_symbol(x, y);
+
+        let pool_manager_signer = &get_pool_account_signer();
+        let pool_cref = &object::create_named_object(pool_manager_signer, pool_address_seed(x, y));
```

Comment:
If we'recreating named objects for the liquidity pool anyway, do we still need the on-chain map? (LiquidityPoolManager)